src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/CoffObjectWriter.cs

```diff
@@ -0,0 +1,1146 @@
+        }
+
+        private sealed class CoffSectionSymbol : CoffSymbolRecord
+        {
+            public uint SizeOfRawData { get; set; }
+            public ushort NumberOfRelocations { get; set; }
+            public ushort NumberOfLineNumbers { get; set; }
+            public uint CheckSum { get; set; }
+            public uint Number { get; set; }
+            public CoffComdatSelect Selection { get; set; }
+
+            private const int RegularSize =
+                sizeof(uint) +   // SizeOfRawData
+                sizeof(ushort) + // NumberOfRelocations
+                sizeof(ushort) + // NumberOfLineNumbers
+                sizeof(uint) +   // CheckSum
+                sizeof(ushort) + // Number
+                sizeof(byte) +   // Selection
+                3;               // Reserved
+
+            private const int BigObjSize = RegularSize + 2;
+
+            public override void Write(Stream stream, CoffStringTable stringTable, bool isBigObj)
+            {
+                Span<byte> buffer = stackalloc byte[isBigObj ? BigObjSize : RegularSize];
+
+                buffer.Clear();
+                BinaryPrimitives.WriteUInt32LittleEndian(buffer, SizeOfRawData);
+                BinaryPrimitives.WriteUInt16LittleEndian(buffer.Slice(4), NumberOfRelocations);
+                BinaryPrimitives.WriteUInt16LittleEndian(buffer.Slice(6), NumberOfLineNumbers);
+                BinaryPrimitives.WriteUInt32LittleEndian(buffer.Slice(8), CheckSum);
+                BinaryPrimitives.WriteUInt16LittleEndian(buffer.Slice(12), (ushort)Number);
+                buffer[14] = (byte)Selection;
+                if (isBigObj)
+                {
+                    BinaryPrimitives.WriteUInt16LittleEndian(buffer.Slice(16), (ushort)(Number >> 16));
+                }
+
+                stream.Write(buffer);
+            }
+        }
+
+        private sealed class CoffStringTable : StringTableBuilder
+        {
+            public new uint Size => (uint)(base.Size + 4);
+
+            public new uint GetStringOffset(string text)
+            {
+                return base.GetStringOffset(text) + 4;
+            }
+
+            public new void Write(FileStream stream)
+            {
+                Span<byte> stringTableSize = stackalloc byte[4];
+                BinaryPrimitives.WriteUInt32LittleEndian(stringTableSize, Size);
+                stream.Write(stringTableSize);
+                base.Write(stream);
+            }
+        }
+
+        /// <summary>
+        /// Checksum algorithm used for COMDAT sections. This is similar to standard
+        /// CRC32 but starts with 0 as initial value instead of ~0.
+        /// </summary>
+        private static class JamCrc32
+        {
+            // NOTE:
+            // This can be generated by Crc32ReflectedTable.Generate(0xEDB88320u);
+            // We embed the pre-generated version since it's small.
+            private static uint[] Table =
+            {
```

Comment:
nit: `readonly` and it can use collection expression:
```c#
private static readonly uint[] s_table =
[
    0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
    ...
];
```

Reply:
I like it, will push it with next batch of changes.

Reply:
Yup, it can be a local `ReadOnlySpan<uint>` in `CalculateChecksum` (e.g. [this](https://github.com/dotnet/runtime/blob/fbf109b6b7f749e899b8859d0eea5ba75b9136ef/src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/WellKnownOids.cs#L118)), and roslyn will put it in data section.

Reply:
Done

Reply:
> (also I think uppercase hex digits are preferred)

Here is a script to bulk transform:

```sh
git checkout objwriter3
git remote add dotnet https://github.com/dotnet/runtime
git fetch dotnet

# only replace in files changed in this PR and still exist
filteredFiles=$(git show --name-only --pretty="" dotnet/main..HEAD |
  while read file; do [ -e "$file" ] && echo "$file"; done)

# before
echo $filterFiles | xargs grep -E '0x[a-f]+' | wc -l

# ssed is gnu-sed on mac used for \U extension, for linux replace ssed with sed
echo $filteredFiles | xargs ssed -i 's/0x\([0-9a-fA-F]*\)/0x\U\1/g'

# after
echo $filterFiles | xargs grep -E '0x[a-f]+' | wc -l
```

Reply:
> (also I think uppercase hex digits are preferred)

Here is a script to bulk transform:

```sh
git checkout objwriter3
git remote add dotnet https://github.com/dotnet/runtime
git fetch dotnet

# only replace in files changed in this PR and still exist
filteredFiles=$(git show --name-only --pretty="" dotnet/main..HEAD |
  while read file; do [ -e "$file" ] && echo "$file"; done)

# before
echo $filterFiles | xargs grep -E '0x[0-9A-F]*[a-f]' | wc -l

# ssed is gnu-sed on mac used for \U extension, for linux replace ssed with sed
echo $filteredFiles | xargs ssed -i 's/0x\([0-9A-F]*[a-f]*\)/0x\U\1/g'

# after
echo $filterFiles | xargs grep -E '0x[0-9A-F]*[a-f]' | wc -l
```

Reply:
Thanks. I'll run the script and check output... and if everything looks okay, push it with the next batch of changes.

Reply:
The code base is actually quite inconsistent in the upper case usage, so the script tried to replace 33 usages in Relocation.cs that were pre-existing. I reverted that part and updated the rest of usages.